### PR TITLE
nrf_wifi: Add lockup recovery stats

### PR DESCRIPTION
**Pull request by kapbh**
Include all lockup recovery stats
under lmac which can be display in
host debug rpu stats.
[[SHEL-2078]](https://nordicsemi.atlassian.net/browse/SHEL-2078)

[SHEL-2078]: https://nordicsemi.atlassian.net/browse/SHEL-2078?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: nrf_wifi/fw_if/umac_if/inc/fw/host_rpu_sys_if.h [hunk 1]
```diff
@@ -630,6 +630,12 @@ struct rpu_lmac_stats {
 	unsigned int scan_abort_complete;
 	/** Number of internal buffer pool null counts */
 	unsigned int internal_buf_pool_null;
+	/** Lockup recovery event*/
+	unsigned int rpuLockupEvent;
+	/** Lockup revovery count*/
+	unsigned int rpuLockupCnt;
+	/** Lockup recovery state*/
+	unsigned int rpuLokcupRecoveryDone;
```
Comment (udaynordic) [inline nrf_wifi/fw_if/umac_if/inc/fw/host_rpu_sys_if.h:638]: We will not be able to modify the existing structure due to ROM firmware. Please add new Command/event to get additional debug data.

**Review by udaynordic — CHANGES_REQUESTED** [review]

File: nrf_wifi/fw_if/umac_if/inc/fmac_api_common.h [hunk 2]
```diff
@@ -358,6 +358,22 @@ enum nrf_wifi_status nrf_wifi_fmac_set_packet_filter(void *dev_ctx, unsigned cha
...
 
+/**
+ * @brief Issue a request to get stats from the RPU.
+ * @param fmac_dev_ctx Pointer to the UMAC IF context for a RPU WLAN device.
+ * @param op_mode RPU operation mode.
+ * @param stats Pointer to memory where the stats are to be copied.
+ *
+ * This function is used to send a command to
+ *	    instruct the firmware to return the current RPU Lockup recovery statistics. The RPU will
+ *	    send the event with the current statistics.
```
Comment (sachinthegreen) [inline nrf_wifi/fw_if/umac_if/inc/fmac_api_common.h:369]: Unnecessary white space at the start of the comments

File: nrf_wifi/fw_if/umac_if/inc/fmac_api_common.h [hunk 3]
```diff
@@ -358,6 +358,22 @@ enum nrf_wifi_status nrf_wifi_fmac_set_packet_filter(void *dev_ctx, unsigned cha
...
+ *
+ * @return Command execution status
+ */
+enum nrf_wifi_status nrf_wifi_fmac_stats_recovery_get(struct nrf_wifi_fmac_dev_ctx *fmac_dev_ctx,
+					     enum rpu_op_mode op_mode,
+					     struct rpu_op_stats *stats);
```
Comment (sachinthegreen) [inline nrf_wifi/fw_if/umac_if/inc/fmac_api_common.h:375]: Indentation

File: nrf_wifi/fw_if/umac_if/inc/fmac_structs_common.h [hunk 4]
```diff
@@ -47,6 +58,8 @@ struct rpu_op_stats {
 	struct rpu_host_stats host;
 	/** Firmware statistics. */
 	struct rpu_fw_stats fw;
+	/** Firmware debug statistics*/
```
Comment (sachinthegreen) [inline nrf_wifi/fw_if/umac_if/inc/fmac_structs_common.h:61]:
```suggestion
	/** Firmware debug statistics */
```

File: nrf_wifi/fw_if/umac_if/src/cmd.c [hunk 5]
```diff
@@ -529,3 +529,44 @@ enum nrf_wifi_status umac_cmd_prog_stats_get(struct nrf_wifi_fmac_dev_ctx *fmac_
 out:
 	return status;
 }
+
+enum nrf_wifi_status umac_cmd_prog_debug_stats_get(struct nrf_wifi_fmac_dev_ctx *fmac_dev_ctx,
+#ifdef CONFIG_NRF700X_RADIO_TEST
+					     int op_mode,
+#endif /* CONFIG_NRF700X_RADIO_TEST */
+					     int stats_type)
```
Comment (sachinthegreen) [inline nrf_wifi/fw_if/umac_if/src/cmd.c:537]: Indentation

File: nrf_wifi/fw_if/umac_if/src/event.c [hunk 6]
```diff
@@ -964,6 +964,46 @@ nrf_wifi_fmac_if_mode_set_event_proc(struct nrf_wifi_fmac_dev_ctx *fmac_dev_ctx,
 }
 #endif
 
+static enum nrf_wifi_status umac_event_stats_debug_process(struct nrf_wifi_fmac_dev_ctx *fmac_dev_ctx,
+						     void *event)
```
Comment (sachinthegreen) [inline nrf_wifi/fw_if/umac_if/src/event.c:968]: Indentation

File: nrf_wifi/fw_if/umac_if/src/event.c [hunk 7]
```diff
@@ -992,6 +1032,10 @@ static enum nrf_wifi_status umac_process_sys_events(struct nrf_wifi_fmac_dev_ctx
 		status = umac_event_stats_process(fmac_dev_ctx,
 						  sys_head);
 		break;
+	case NRF_WIFI_EVENT_DEBUG_STATS:
+		status = umac_event_stats_debug_process(fmac_dev_ctx,
+								sys_head);
```
Comment (sachinthegreen) [inline nrf_wifi/fw_if/umac_if/src/event.c:1037]: Indentation

File: nrf_wifi/fw_if/umac_if/src/fmac_api_common.c [hunk 8]
```diff
@@ -1248,3 +1248,61 @@ enum nrf_wifi_status nrf_wifi_fmac_set_packet_filter(void *dev_ctx, unsigned cha
 	return status;
 }
 #endif /* CONFIG_NRF700X_RAW_DATA_RX || CONFIG_NRF700X_PROMISC_DATA_RX */
+
+enum nrf_wifi_status nrf_wifi_fmac_stats_recovery_get(struct nrf_wifi_fmac_dev_ctx *fmac_dev_ctx,
+					     enum rpu_op_mode op_mode,
+					     struct rpu_op_stats *stats)
```
Comment (sachinthegreen) [inline nrf_wifi/fw_if/umac_if/src/fmac_api_common.c:1254]: Indentation

File: nrf_wifi/fw_if/umac_if/inc/fmac_structs_common.h [hunk 9]
```diff
@@ -38,6 +38,17 @@ struct rpu_host_stats {
...
+/**
+ * @brief Structure to hold firmware debug statistics.
+ *
+ */
+struct rpu_fw_debug_stats{
+	/** RPU hardware lockup event detection count */
+	unsigned int rpu_hw_lockup_count;
+	/** RPU hardware lockup recovery completed count */
+	unsigned int rpu_hw_lockup_recovery_done;
+};
```
Comment (sachinthegreen) [inline nrf_wifi/fw_if/umac_if/inc/fmac_structs_common.h:50]: Why does this structure need to be defined in the driver header file ? Shouldn't this be picked up from the FW shared header file?

File: nrf_wifi/fw_if/umac_if/src/fmac_api_common.c [hunk 10]
```diff
@@ -1248,3 +1248,61 @@ enum nrf_wifi_status nrf_wifi_fmac_set_packet_filter(void *dev_ctx, unsigned cha
...
+		}
+
+		fmac_dev_ctx->stats_req = true;
+		fmac_dev_ctx->fw_debug = &stats->fw_debug;
+		
+		status = umac_cmd_prog_debug_stats_get(fmac_dev_ctx,
+		#ifdef CONFIG_NRF700X_RADIO_TEST
+						 op_mode,
+		#endif /* CONFIG_NRF700X_RADIO_TEST */
+						 stats_type);
```
Comment (sachinthegreen) [inline nrf_wifi/fw_if/umac_if/src/fmac_api_common.c:1283]: Indentation

**Review by sachinthegreen — CHANGES_REQUESTED** [review]

File: nrf_wifi/fw_if/umac_if/inc/fmac_api_common.h (hunk 2 above)
Comment (kapbh) [reply nrf_wifi/fw_if/umac_if/inc/fmac_api_common.h:369]: It's part of format. Every comments have same white space.

File: nrf_wifi/fw_if/umac_if/inc/fmac_structs_common.h (hunk 9 above)
Comment (kapbh) [reply nrf_wifi/fw_if/umac_if/inc/fmac_structs_common.h:50]: There is no structure available to store the value. They have defined the variables with event structure. So need to store the value.

File: nrf_wifi/fw_if/umac_if/src/fmac_api_common.c [hunk 11]
```diff
@@ -1281,3 +1281,61 @@ enum nrf_wifi_status nrf_wifi_fmac_set_packet_filter(void *dev_ctx, unsigned cha
...
+
+enum nrf_wifi_status nrf_wifi_fmac_stats_recovery_get(struct nrf_wifi_fmac_dev_ctx *fmac_dev_ctx,
+						      enum rpu_op_mode op_mode,
+						      struct rpu_op_stats *stats)
+{
+	enum nrf_wifi_status status = NRF_WIFI_STATUS_FAIL;
+	unsigned char count = 0;
+	int stats_type;
+
+	#ifdef CONFIG_NRF700X_RADIO_TEST
```
Comment (krish2718) [inline nrf_wifi/fw_if/umac_if/src/fmac_api_common.c:1293]: These shouldn't be aligned.

**Review by krish2718 — APPROVED** [review]

File: nrf_wifi/fw_if/umac_if/inc/fmac_api_common.h [hunk 12]
```diff
@@ -358,6 +358,23 @@ enum nrf_wifi_status nrf_wifi_fmac_set_packet_filter(void *dev_ctx, unsigned cha
...
+ * @param op_mode RPU operation mode.
+ * @param stats Pointer to memory where the stats are to be copied.
+ *
+ * This function is used to send a command to
+ * instruct the firmware to return the current RPU Lockup recovery statistics. The RPU will
+ * send the event with the current statistics.
+ *
+ * @return Command execution status
+ */
+enum nrf_wifi_status nrf_wifi_fmac_stats_recovery_get(struct nrf_wifi_fmac_dev_ctx *fmac_dev_ctx,
```
Comment (sachinthegreen) [inline nrf_wifi/fw_if/umac_if/inc/fmac_api_common.h:374]:
```suggestion
enum nrf_wifi_status nrf_wifi_fmac_dbg_stats_get(struct nrf_wifi_fmac_dev_ctx *fmac_dev_ctx,
```

**Review by sachinthegreen — CHANGES_REQUESTED** [review]

**Review by sachinthegreen — CHANGES_REQUESTED** [review]
As discussed offline with you and @srinivas1201, this is better handled in the firmware so that user exposed code does not become too complicated and also leads to user friendly documentation.

**Comment by kapbh** [comment]
LMAC recovery stats are now part of LMAC stats, no need to have new command/events and structure. Closing this PR.

**Pull request closed without merging**